src/meta/meta_turtle.c

```diff
@@ -81,12 +93,88 @@ __metadata_load_hot_backup(WT_SESSION_IMPL *session)
+         * lexicographically sorted, we can use a two pointer method to check if the current metadata
+         * entry needs to be dropped. If the metadata entry doesn't exist in the target uri list,
+         * append the table name to the partial backup list.
+         */
+        metadata_key = (char *)key->data;
+        if (F_ISSET(conn, WT_CONN_BACKUP_PARTIAL_RESTORE) &&
+          WT_PREFIX_MATCH(metadata_key, "table:")) {
+            /*
+             * The target uri will be the deciding factor if a specific metadata table entry needs
+             * to be dropped. The entry in the target uri aims to be either equal or greater than 
+             * the current metadata entry. This allows us to handle two cases to decide if
+             * we need to append the metadata key to the backup list.
+             * 
+             * 1. If the metadata key is less than the current target uri, this means that the metadata
+             * entry needs to be dropped.
+             * 2. If the target uri is NULL, meaning that we have parsed through all of the entries
+             * in the target list. All following metadata entries needs to be dropped.
+             */
+            if (target_uris[cnt] == NULL || strcmp(metadata_key, target_uris[cnt]) < 0) {
+                WT_ERR(__wt_realloc_def(session, &allocated_name, slot + 2, &partial_backup_names));
+                p = &partial_backup_names[slot];
+                p[0] = p[1] = NULL;
+
+                WT_ERR(
+                  __wt_strndup(session, (char *)key->data, key->size, &partial_backup_names[slot]));
+                slot++;
+            } else if (strcmp(metadata_key, target_uris[cnt]) == 0)
+                cnt++;
+        }
+        /*
+         * In the case of partial backup restore, add the entry to the metadata even if the file
+         * doesn't exist so that we can correctly drop all related entries via the schema code
+         * later.
+         */
         WT_ERR(__wt_metadata_update(session, key->data, value->data));
     }
 
-    F_SET(S2C(session), WT_CONN_WAS_BACKUP);
+    F_SET(conn, WT_CONN_WAS_BACKUP);
+    if (F_ISSET(conn, WT_CONN_BACKUP_PARTIAL_RESTORE) && partial_backup_names != NULL) {
+        WT_ERR(__wt_calloc_def(session, slot + 1, &conn->partial_backup_remove_ids));
+        /*
+         * Parse through the partial backup list and attempt to clean up all metadata references
+         * relating to the file. To do so, perform a schema drop operation on the table to cleanly
+         * remove all linked references. At the same time generate a list of btree ids to be used in
+         * recovery to truncate all the history store records.
+         */
+        for (slot = 0; partial_backup_names[slot] != NULL; ++slot) {
+            tablename = partial_backup_names[slot];
+            WT_PREFIX_SKIP_REQUIRED(session, tablename, "table:");
+            len = strlen("file:") + strlen(tablename) + strlen(".wt") + 1;
```

Comment:
Since we know that mongodb table names are generally consistent and mostly fixed length, it would be better to keep track of the max length seen so far and only reallocate if the new length exceeds the already allocated one rather than allocate and free every time through here.